File: packages/ai/models/game_state/feat.py
from datetime import datetime
from pydantic import BaseModel


class Feat(BaseModel):
    name: str
    cooldown: int
    cooldownEndTime: int
    castingDuration: int
    isCasting: bool
    isReady: bool
    category: str

    def time_remaining(self) -> int:
        now = int(datetime.utcnow().now().timestamp() * 1000)
        remaining = self.cooldownEndTime - now
        return remaining // 1000 if remaining > 0 else 0

    def __repr__(self):
        if self.isReady:
            status = "ready"
        else:
            status = f"ready in {self.time_remaining()}s"

        return f"<Feat: name={self.name}, category={self.category}, status={status}>"

File: packages/server/src/game/modules/feats/feat.ts
import { Schema, type } from "@colyseus/schema";
import { Entity } from "../../entities/entity";

export class Feat extends Schema {
  @type("string")
  name: string;

  entity: Entity;

  passive: boolean = false;

  @type("number")
  cooldown: number = 0;

  @type("number")
  cooldownEndTime: number = 0;

  @type("number")
  castingDuration: number = 0;

  @type("boolean")
  isCasting: boolean = false;

  @type("boolean")
  isReady: boolean = true;

  manaCost: number = 0;

  @type("string")
  category: string = "offensive";

  ranged: boolean = true;

  constructor(name: string, entity: Entity) {
    super();
    this.name = name;
    this.entity = entity;
  }

  use() {
    if (!this.isValid()) return;

    this.effect();

    if (this.cooldown > 0) {
      this.cooldownEndTime = Date.now() + this.cooldown * 1000;
      this.isReady = false;
    }
  }

  update() {
    if (this.isCasting || this.isReady) return;

    if (Date.now() >= this.cooldownEndTime) {
      this.isReady = true;
    }
  }

  effect() {}

  execute() {}

  isValid(): boolean {
    return this.isReady;
  }
}
